validator/sawtooth_validator/state/batch_tracker.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2017 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# -----------------------------------------------------------------------------
+
+import abc
+from threading import RLock
+from sawtooth_validator.protobuf import client_pb2
+from sawtooth_validator.journal.block_store import CommittedBlocksObserver
+from sawtooth_validator.journal.completer import PendingBatchObserver
+
+
+class BatchTracker(CommittedBlocksObserver, PendingBatchObserver):
+    """Tracks batch statuses for this local validator, allowing interested
+    components to check where a batch is in the validation process. It should
+    only be relied on for batches submitted locally, and is not persisted
+    after restart.
+
+    When a batch moves from one component to another, the appropriate notify
+    method should be called in the appropriate component:
+        * notify_committed
+        * notify_pending
+    """
+    def __init__(self):
+        self._committed = set()
```

Comment:
They are unbounded. `_committed` will include the `header_signature` of every block in the blockchain, and `_pending` will include every submitted batch _not_ in the blockchain.

In a future PR, a timed cache for invalid batches will be included, which means every submitted batch will end up in either `_committed` or `_invalid` and the invalid ones will expire. However, until that PR is merged, one could spam invalid batches and clog up the `_pending` list. I will make `_pending` a timed cache until that next PR comes in.